tool/src/org/antlr/v4/codegen/Target.java

```diff
@@ -329,21 +343,70 @@ public String encodeIntAsCharEscape(int v) {
 			throw new IllegalArgumentException(String.format("Cannot encode the specified value: %d", v));
 		}
 
-		if ( v < targetCharValueEscape.length && targetCharValueEscape[v] != null) {
-			return targetCharValueEscape[v];
+		String language = getLanguage();
+		if (language.equals("Go") || language.equals("Cpp")) {
+			return Integer.toString(v);
+		}
+
+		char c = (char)v;
+		String escaped = getTargetCharValueEscape().get(c);
+		if (escaped != null) {
+			isPreviousOctal = false;
+			return escaped;
 		}
 
-		if (v >= 0x20 && v < 127 && (!Character.isDigit(v) || v == '8' || v == '9')) {
-			return String.valueOf((char)v);
+		switch (Character.getType(c))
+		{
+			case Character.CONTROL:
+			case Character.LINE_SEPARATOR:
+			case Character.PARAGRAPH_SEPARATOR:
+				return escapeChar(v);
+			default:
+				if (v == 0xfffe) {
+					return escapeChar(v);
+				}
+
+				if (isPreviousOctal) {
+					char upperBound = language.equals("PHP") ? '9' : '7';
+					if (c >= '0' && c <= upperBound) {
+						return escapeChar(v);
+					}
+				}
+
+				isPreviousOctal = false;
+				return String.valueOf(c);
 		}
+	}
 
-		if ( v<=127 ) {
-			String oct = Integer.toOctalString(v);
-			return "\\"+ oct;
+	private String escapeChar(int v) {
+		String language = getLanguage();
+
+		boolean isPhp = language.equals("PHP");
+		boolean supportsOctalEncoding = language.equals("Java")
+				|| language.equals("Python2")
+				|| language.equals("Python3")
+				|| isPhp;
+		if (supportsOctalEncoding && v <= (isPhp ? 127 : 255)) {
+			isPreviousOctal = true;
+			return String.format("\\%o", v);
+		} else {
+			isPreviousOctal = false;
 		}
 
-		String hex = Integer.toHexString(v|0x10000).substring(1,5);
-		return "\\u"+hex;
+		switch (language) {
+			default:
+			case "Java":
+			case "JavaScript":
+			case "Python2":
+			case "Python3":
+				return String.format("\\u%04x", v);
+			case "CSharp":
+				return String.format("\\x%X", v);
+			case "Dart":
+			case "PHP":
+			case "Swift":
+				return String.format("\\u{%X}", v);
```

Comment:
I suggest adding `StandardTarget` eventually (we were discussing the question some time ago) to minimize the risk of misprinting in string literal:

```java
class StandardTarget {
    public final static String Java = "Java";
    ...
}
```

or add constants to each target:

```java
class JavaTarget {
    public final static String Key = "Java";
}
```

Those constants are used over the repository in many places.

Reply:
If the Constants are used a lot, I can see creating such a mechanism. For now, let's focus on just the changes directly in the serialization.